Mathlib/Tactic/Linter/TextBased.lean

```diff
@@ -0,0 +1,196 @@
+    if !(copy.startsWith "Copyright (c) 20" && copy.endsWith ". All rights reserved.") then
+      output := output.push (StyleError.copyright "Copyright line is malformed", 2)
+  -- The second line should be standard.
+  let expected_second_line := "Released under Apache 2.0 license as described in the file LICENSE."
+  if start.get? 2 != some expected_second_line then
+    output := output.push (StyleError.copyright
+      s!"Second line should be \"{expected_second_line}\"", 3)
+  -- The third line should contain authors.
+  if let some line := start.get? 3 then
+    if !line.containsSubstr "Author" then
+      output := output.push (StyleError.copyright
+        "The third line should describe the file's main authors", 4)
+    else
+      -- If it does, we check the authors line is formatted correctly.
+      if !is_correct_authors_line line then
+        output := output.push (StyleError.authors, 4)
+  return output
+
+
+/-- Whether a collection of lines consists *only* of imports:
+in practice, this means it's an imports-only file and exempt from almost all linting. -/
+def is_imports_only_file (lines : Array String) : Bool :=
+  -- The Python version also excluded comments: since the import-only files are
+  -- automatically generated and don't contains comments, this is in fact not necessary.
+  -- XXX: also implement parsing of multi-line comments.
+  lines.all (fun line ↦ line.startsWith "import " || line == "" || line.startsWith "-- ")
+
+end
+
+/-- All text-based linters registered in this file. -/
+def all_linters : Array LinterCore := Array.mk
+  [copyright_header]
+
+/-- Read a file, apply all text-based linters and return the formatted errors.
+
+Return `true` if there were new errors (and `false` otherwise). -/
+def lint_file (path : System.FilePath) : IO Bool := do
+  let lines ← IO.FS.lines path
+  -- We don't need to run any checks on imports-only files.
+  -- NB. The Python script used to still run a few linters; this is in fact not necessary.
+  if is_imports_only_file lines then
+    return false
+  let all_output := (Array.map (fun lint ↦
+    (Array.map (fun (e, n) ↦ ErrorContext.mk e n path)) (lint lines))) all_linters
+  -- XXX: this list is currently not sorted: for github, that's probably fine
+  let errors := Array.flatten all_output
+  formatErrors errors
+  return errors.size > 0
+
+/-- Lint all files referenced in a given import-only file.
+Return the number of files which had new style errors. -/
+def lint_all_files (path : System.FilePath) : IO UInt32 := do
+  -- Read all module names in Mathlib from the file at `path`.
+  let allModules ← IO.FS.lines path
+  let mut number_error_files := 0
+  for module in allModules do
+    let module := module.stripPrefix "import "
+    -- Exclude `Archive.Sensitivity` and `Mathlib.Tactic.Linter` for now.
+    -- FUTURE: replace this by proper parsing of style exceptions.
+    if #["Archive.Sensitivity", "Mathlib.Tactic.Linter"].contains module then
+      continue
```

Comment:
I'd argue that having support for the style exceptions in nolints is a prerequisite for rewriting these linters.

Reply:
In principle, I wholeheartedly agree: indeed, I have implemented parsing style exceptions and updating their exceptions file. I am happy to add it to this PR, if you prefer. I origininally did not, to makes this PR land faster.

I think parsing style exceptions or updating them is essential for rewriting *this* linter:
- there's no reason new copyright exceptions should get added
(This is different from e.g. the large file linter `ERR_NUM_LIN`, where exceptions are an accepted transient state.)
- all existing copyright headers pass the Python lint
- these hard-coded exceptions are one-off cases where the rewritten linter is too strict
Both of them are not hard to fix (but make the code more complicated): do you really care to get this fixed now?

Reply:
Fair enough, I have implemented this now.